contracts/finality/src/state/allowlist.rs

```diff
@@ -1,49 +1,194 @@
-    Map::new("allowed_finality_providers");
+// Storage key constants
+const ALLOWED_FINALITY_PROVIDERS_KEY: &str = "allowed_finality_providers";
+const ALLOWED_FINALITY_PROVIDERS_CHECKPOINTS_KEY: &str = "allowed_finality_providers__checkpoints";
+const ALLOWED_FINALITY_PROVIDERS_CHANGELOG_KEY: &str = "allowed_finality_providers__changelog";
 
-/// Check if a finality provider is in the allowlist
+/// SnapshotItem of allowed finality provider BTC public keys stored as a HashSet
+pub(crate) const ALLOWED_FINALITY_PROVIDERS: SnapshotItem<HashSet<Vec<u8>>> = SnapshotItem::new(
+    ALLOWED_FINALITY_PROVIDERS_KEY,
+    ALLOWED_FINALITY_PROVIDERS_CHECKPOINTS_KEY,
+    ALLOWED_FINALITY_PROVIDERS_CHANGELOG_KEY,
+    Strategy::EveryBlock,
+);
+
+/// Check if a finality provider is in the allowlist (at current Babylon height)
 pub fn ensure_fp_in_allowlist(
     storage: &dyn Storage,
     fp_btc_pk_bytes: &[u8],
 ) -> Result<(), ContractError> {
-    ALLOWED_FINALITY_PROVIDERS
-        .has(storage, fp_btc_pk_bytes)
-        .then_some(())
-        .ok_or(ContractError::FinalityProviderNotAllowed(hex::encode(
+    let fp_set = ALLOWED_FINALITY_PROVIDERS
+        .may_load(storage)?
+        .unwrap_or_default();
+
+    if fp_set.contains(fp_btc_pk_bytes) {
+        Ok(())
+    } else {
+        Err(ContractError::FinalityProviderNotAllowed(hex::encode(
+            fp_btc_pk_bytes,
+        )))
+    }
+}
+
+/// Check if a finality provider was in the allowlist at a specific Babylon height
+pub fn ensure_fp_in_allowlist_at_height(
+    storage: &dyn Storage,
+    fp_btc_pk_bytes: &[u8],
+    babylon_height: u64,
+) -> Result<(), ContractError> {
+    let fp_set = ALLOWED_FINALITY_PROVIDERS
+        .may_load_at_height(storage, babylon_height)
+        .map_err(ContractError::StdError)?
+        .unwrap_or_default();
+
+    if fp_set.contains(fp_btc_pk_bytes) {
+        Ok(())
+    } else {
+        Err(ContractError::FinalityProviderNotAllowed(hex::encode(
             fp_btc_pk_bytes,
         )))
+    }
 }
 
-/// Add a finality provider to the allowlist
+/// Add a finality provider to the allowlist at a specific Babylon height
 pub fn add_finality_provider_to_allowlist(
     storage: &mut dyn Storage,
     fp_btc_pk_bytes: &[u8],
+    babylon_height: u64,
 ) -> Result<(), ContractError> {
+    let mut fp_set = ALLOWED_FINALITY_PROVIDERS
+        .may_load(storage)?
+        .unwrap_or_default();
+    fp_set.insert(fp_btc_pk_bytes.to_vec());
     ALLOWED_FINALITY_PROVIDERS
-        .save(storage, fp_btc_pk_bytes, &())
+        .save(storage, &fp_set, babylon_height)
         .map_err(Into::into)
 }
 
-/// Remove a finality provider from the allowlist
+/// Remove a finality provider from the allowlist at a specific Babylon height
 pub fn remove_finality_provider_from_allowlist(
     storage: &mut dyn Storage,
     fp_btc_pk_bytes: &[u8],
+    babylon_height: u64,
 ) -> Result<(), ContractError> {
```

Comment:
I think we can take a list of FP PKs in this function, such that we remove all given FPs in the hashset, and write back to DB. This saves `x` snapshots where `x` is the length of the FP PK list. Same comment applies to addition

Reply:
But if the admin wants to add/remove `X` FPs in a single call, then the current implementation might create `X` snapshots; meanwhile if we modify all `X` FPs in the hashset then write back, then there is only 1 snapshot.